go/graphd/io.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2018 Google Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The io portion of the graphd package facilitates IO operations against a graphd database.
+
+package graphd
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+)
+
+// readReponses returns a Response pointer slice read from an established connection to a graphd
+// database and a nil error on success.  On any failure to read a response, a Response pointer
+// slice containing zero-value Response pointers for the given failures, and the last encountered
+// error are returned.  readResponses must be called with g.conn locked.
+// TODO: Would it be useful to return all encountered errors?  Any chance after a failure that
+//       we'll block on the following call to ReadString()?
+func (g *graphd) readResponses(reqsNum int) ([]*Response, error) {
+	var retErr error
+	var resSlice []*Response
+
+	reader := bufio.NewReader(g.conn.netConn)
+
+	for i := 0; i < reqsNum; i++ {
+		str, err := reader.ReadString('\n')
+		if err != nil {
+			resSlice = append(resSlice, NewResponse(""))
+			retErr = err
+		} else {
+			resSlice = append(resSlice, NewResponse(str))
+		}
+	}
+
+	return resSlice, retErr
+}
+
+// Query attempts to send the Request to the graphd database to which this instance of the library
+// is connected.  In the case of more than one Request, the requests are joined and sent as one.
+// If no connection is currently present, or if the established connection is stale, Query will
+// trigger a Redial.  A Response pointer slice containing responses from the graphd database (failed
+// responses are zero-value value Response pointers) is returned along with an error code.  In the
+// case of any failed responses, the returned error code will contain the last encountered error.
+// Query locks the connection, allowing one thread to Query at a time.
```

Comment:
Locking the connection i think is fine for right now, but the thing to consider next is how to interleave queries - graphd can process > 1 query at a time, so maybe the next step here is to figure out how multiplexing should work (we talked about that a little i think?)

Reply:
Well the lock is to serialize the send/receive for a given thread.  However, in the case where Query gets more than one Request instance in the parameter list, it joins the requests together and sends them to graphd as one.  It then waits for N responses.  If we want to interleave further (ie. break serialization), we'd have to go with the original idea of a separate receiver thread, etc.

Anyway, I'll ping you offline in the morning and we can discuss this one.